docs/markdown/Contributing.md

```diff
@@ -163,28 +163,136 @@ Any tests that require more thorough analysis, such as checking that certain
 compiler arguments can be found in the command line or that the generated
 pkg-config files actually work should be done with a unit test.
 
-The following files in the test's source root are consulted, if they exist:
-
-* `installed_files.txt` lists the files which are expected to be installed.
-Various constructs containing `?` are used to indicate platform specific
-filename variations (e.g. `?so` represents the platform appropriate suffix for a
-shared library)
-
-* `setup_env.json` contains a dictionary which specifies additional
-environment variables to be set during the configure step of the test. `@ROOT@`
-is replaced with the absolute path of the source directory.
+Additionally:
 
 * `crossfile.ini` and `nativefile.ini` are passed to the configure step with
 `--cross-file` and `--native-file` options, respectively.
 
-Additionally:
-
 * `mlog.cmd_ci_include()` can be called from anywhere inside meson to capture the
 contents of an additional file into the CI log on failure.
 
 Projects needed by unit tests are in the `test cases/unit`
 subdirectory. They are not run as part of `./run_project_tests.py`.
 
+#### Configuring project tests
+
+The (optional) `test.json` file, in the root of a test case, is used
+for configuring the test. All of the following root entries in the `test.json`
+are independent of each other and can be combined as needed.
+
+Exanple `test.json`:
+
+```json
+{
+  "env": {
+    "VAR": "VAL"
+  },
+  "installed": [
+    { "type": "exe", "file": "usr/bin/testexe" },
+    { "type": "pdb", "file": "usr/bin/testexe" },
+  ],
+  "matrix": {
+    "options": {
+      "opt1": [
+        { "val": "abc"   },
+        { "val": "qwert" },
+        { "val": "bad"   }
+      ],
+      "opt2": [
+        { "val": null    },
+        { "val": "true"  },
+        { "val": "false" },
+      ]
+    },
+    "exclude": [
+      { "opt1": "qwert", "opt2": "false" },
+      { "opt1": "bad"                    }
+    ]
+  }
+}
+```
+
+##### env
+
+The `env` key contains a dictionary which specifies additional
+environment variables to be set during the configure step of the test. `@ROOT@`
```

Comment:
Maybe it would be better to use `@SOURCE_DIR@` like we do in the meson language? I don't know, I'm just thinking out loud.